lib/vrf.c

```diff
@@ -384,54 +384,74 @@ static void vrf_hash_bitmap_free(void *data)
 	XFREE(MTYPE_VRF_BITMAP, bit);
 }
 
-vrf_bitmap_t vrf_bitmap_init(void)
+void vrf_bitmap_init(vrf_bitmap_t *pbmap)
 {
-	return hash_create_size(32, vrf_hash_bitmap_key, vrf_hash_bitmap_cmp,
-				"VRF BIT HASH");
+	*pbmap = NULL;
 }
 
-void vrf_bitmap_free(vrf_bitmap_t bmap)
+void vrf_bitmap_free(vrf_bitmap_t *pbmap)
 {
-	struct hash *vrf_hash = bmap;
+	struct hash *vrf_hash;
+
+	if (!*pbmap)
+		return;
+
+	vrf_hash = *pbmap;
 
 	hash_clean_and_free(&vrf_hash, vrf_hash_bitmap_free);
 }
 
-void vrf_bitmap_set(vrf_bitmap_t bmap, vrf_id_t vrf_id)
+void vrf_bitmap_set(vrf_bitmap_t *pbmap, vrf_id_t vrf_id)
 {
 	struct vrf_bit_set lookup = { .vrf_id = vrf_id };
-	struct hash *vrf_hash = bmap;
+	struct hash *vrf_hash;
 	struct vrf_bit_set *bit;
 
-	if (vrf_hash == NULL || vrf_id == VRF_UNKNOWN)
+	if (vrf_id == VRF_UNKNOWN)
 		return;
 
+	if (!*pbmap)
+		*pbmap = vrf_hash =
+			hash_create_size(2, vrf_hash_bitmap_key,
+					 vrf_hash_bitmap_cmp, "VRF BIT HASH");
+	else
+		vrf_hash = *pbmap;
+
 	bit = hash_get(vrf_hash, &lookup, vrf_hash_bitmap_alloc);
 	bit->set = true;
 }
 
-void vrf_bitmap_unset(vrf_bitmap_t bmap, vrf_id_t vrf_id)
+void vrf_bitmap_unset(vrf_bitmap_t *pbmap, vrf_id_t vrf_id)
 {
 	struct vrf_bit_set lookup = { .vrf_id = vrf_id };
-	struct hash *vrf_hash = bmap;
+	struct hash *vrf_hash;
 	struct vrf_bit_set *bit;
 
-	if (vrf_hash == NULL || vrf_id == VRF_UNKNOWN)
+	if (vrf_id == VRF_UNKNOWN)
 		return;
 
+	/*
+	 * If the hash is not created then unsetting is unnecessary
+	 */
+	if (!*pbmap)
+		return;
+
+	vrf_hash = *pbmap;
+
 	bit = hash_get(vrf_hash, &lookup, vrf_hash_bitmap_alloc);
 	bit->set = false;
 }
 
-int vrf_bitmap_check(vrf_bitmap_t bmap, vrf_id_t vrf_id)
+int vrf_bitmap_check(vrf_bitmap_t *pbmap, vrf_id_t vrf_id)
```

Comment:
to make it extremely consistent imo

Reply:
well, ok, but ... it makes the impact of the change meaningfully greater. when I see the pointer-to-pointer arg I think that the called api may ... make a change, but that's not the case here. I guess that's setting off some dissonance for me.

Reply:
It should have a `const`, but I am in favor of in fact making it a pointer. That way it behaves like a container data structure — all of those always take pointers, even if some read access could work with a by-value parameter.

Having it as a pointer also gives us more freedom to do fancy things with the underlying data structure.

But, as a final caveat, I'd like to move `vrf_bitmap_check` from `vrf.c` to `vrf.h` and make it a `static inline`, exactly to eliminate the cost of it taking a pointer argument. Particularly since `vrf_bitmap_check` is used quite a bit in hot paths.

P.S.: there's always the `[1]` array trick if we want to avoid adding `&` to call sites.